### PR TITLE
Only upgrade :latest casks when --greedy and the cask has been updated

### LINKED ISSUE #11513: Don't run update on casks with `:latest` when the artifacts haven't changed
### Provide a detailed description of the proposed feature

When running `brew upgrade --greedy`, if the artifacts (i.e. files to install) for a cask using `version :latest` has not changed since the last successful install. An example of how this might be done is to compare the hash of the artifacts before and after the download and only installing if they don't match.

### What is the motivation for the feature?

Installing casks that use `version :latest` each time `brew upgrade --greedy` when the artifacts is unchanged is unnecessary, time consuming, resource consuming, as well as potentially interrupting the user's workflow if they relaunch the application, thinking there's new updates.

### How will the feature be relevant to at least 90% of Homebrew users?

Prevent issues stated in "motivation"

### What alternatives to the feature have been considered?

- Installing casks that use `version :latest` each time `brew upgrade --greedy` when the artifacts is unchanged - see "motivation"
- Checking each individual application's website if a new version is released and then updating only those who have a new release - absurdly inconvenient

### PR DESCRIPTION
**Pull request by apainintheneck**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

This PR is related to issue #11513 where `version :latest` casks would always be reinstalled when the `--greedy` or `--greed-latest` flags were passed to `brew upgrade` regardless of whether or not the casks had been updated.

Basically the idea is to store a sha256 hash of the original download file every time a `version :latest` cask is installed. Then, before upgrading casks `cask#outdated?` is queried which internally compares the new and old sha of `version :latest` casks to see if they should be updated. As far as I can see, the first download used to check the sha should be cached so it shouldn't need to be downloaded again before installing it.

The sha is stored in a file called `LATEST_DOWNLOAD_SHA256` in the `.metadata/` cask subdirectory. If there is no previous sha stored, it just automatically assumes the cask should be updated.

Fixes #11513.

### CODE REVIEW

File: Library/Homebrew/cask/cask.rb [hunk 1]
```diff
@@ -127,6 +127,23 @@ def config_path
...
+    def download_sha_path
+      metadata_main_container_path/"LATEST_DOWNLOAD_SHA256"
+    end
+
+    def new_download_sha
+      require "cask/installer"
+
+      @new_download_sha ||= Installer.new(
+        self, verify_download_integrity: false
+      ).download(quiet: true).instance_eval { |x| Digest::SHA256.file(x).hexdigest }
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cask/cask.rb:139]:
```suggestion
      @new_download_sha ||= Installer.new(
        self, verify_download_integrity: false
      ).download(quiet: true)
       .instance_eval { |x| Digest::SHA256.file(x).hexdigest }
```

File: Library/Homebrew/cask/cask.rb [hunk 2]
```diff
@@ -140,21 +157,21 @@ def outdated_versions(greedy: false, greedy_latest: false, greedy_auto_updates:
...
         version
       end
 
+      if greedy || greedy_latest || (greedy_auto_updates && auto_updates)
+        if latest_version.latest?
+          return outdated_download_sha? ? versions : []
+        end
+      elsif auto_updates
+        return []
+      end
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cask/cask.rb:173]:
```suggestion
      if greedy || greedy_latest || (greedy_auto_updates && auto_updates)
        if latest_version.latest?
          if outdated_download_sha?
            return versions
          else
            return []
          end
        end
      elsif auto_updates
        return []
      end
```

**Review by MikeMcQuaid — COMMENTED** [review]
Looks good so far!

Would using an `ETag`, HTTP `HEAD` and/or on-demand SHA hashing allow simplifying any logic?

File: Library/Homebrew/cask/cask.rb (hunk 2 above)
Comment (apainintheneck) [reply Library/Homebrew/cask/cask.rb:173]: Unfortunately, this doesn't pass the linter. We could use a guard clause but that seems even uglier than the ternary operator imo. Which do you prefer?

```ruby
# before
return outdated_download_sha? ? versions : []

# after
return versions if outdated_download_sha?

return []
```

File: Library/Homebrew/cask/cask.rb (hunk 1 above)
Comment (apainintheneck) [reply Library/Homebrew/cask/cask.rb:139]: I think I'll just line up `.new`, `.download` and `.instance_eval` vertically instead.

**Comment by apainintheneck** [comment]
> Would using an `ETag`, HTTP `HEAD` and/or on-demand SHA hashing allow simplifying any logic?

Maybe? I'm not super familiar with `ETag`s to be honest but being able to see if files were changed at an address does sound useful. Plus, I'm not sure how HTTP `HEAD` would be helpful here. Did you something specific in mind?

I don't think adding any of those things would simplify the logic per se but it could help reduce unnecessary downloads and shas which I think would be the biggest bottlenecks here. As far as I can see, a lot of that work is already being done in the download method where it defaults to the cached download if the remote one hasn't been updated (now that I write this maybe something like `ETag`s are being used internally by the downloader to verify that).

With the current code the following things need to be checked when trying to verify if a `:latest version` cask is outdated.
1. No previous sha -> Outdated
   - No download or extra work
3. Compare previous sha to new sha. 
   - They match -> Up to date
   - They don't match -> Outdated
      - The downside of this approach is that the new sha ALWAYS get calculated even if we use the same cached download file.

Would it be possible to just use the `ETag` assuming it is unique enough to verify that the resource is newer than the last install? The thing I don't know off the top of my head is whether or not that those are used universally enough to be able to count on them. They seem to be optional and there is no agreed upon way to generate them.

I think the next thing I'll do is look at how the downloader decides when to use the cached download. This could in theory be used to decide when casks need updating.

**Comment by apainintheneck** [comment]
Okay, I took a look around and I think that HTTP `HEAD` could be used to prevent unnecessarily downloading most files but not all. Here are the problems with using it to determine if the file in question is new.
1. Not all URLs support `HEAD` requests.
2. Those that do don't always include the `Etag` and/or `Last-Modified` elements needed to check the "freshness" of the file.

So in conclusion, we'd still need to use hashing as a backup way to make sure we aren't reinstalling the same cask when upgrading unless we're fine with a 90% solution (that's just an estimate based on looking at casks in the default tap).

It also would mean adding a method/class for retrieving and checking `HEAD` requests and all of that download information like the header elements and hash would need to be stored somewhere, generated and updated when installing `version :latest` casks. I could try and build that by using `CurlDownloadStrategy` as a guide. I'm honestly not sure how involved it would be though.

**Comment by apainintheneck** [comment]
I might be wrong but it seems like at least the `CurlDownloadStrategy` checks if it should use the cached download by making another `GET` request. This means that it will download the same file multiple times which is not ideal.

I believe the download still gets cached though so one solution would be to store that pathname in the cask object instead of the installer. Then, if a cask uses an installer to download a file twice during the same session, it will default to the cached download.

Also, is HTTP `HEAD` used anywhere in the codebase currently?

File: Library/Homebrew/cask/cask.rb (hunk 2 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/cask/cask.rb:173]: # after
return versions if outdated_download_sha?
return []

This seems nice!

**Comment by MikeMcQuaid** [comment]
> I might be wrong but it seems like at least the `CurlDownloadStrategy` checks if it should use the cached download by making another `GET` request. This means that it will download the same file multiple times which is not ideal.

I think it asks `curl` to resume the existing download with possible possibly making one of these a no-op?

> Also, is HTTP `HEAD` used anywhere in the codebase currently?

Unsure! To be clear: was just spitballing enough and I'm happy for this to be merged as-is when you are 👍🏻

**Comment by apainintheneck** [comment]
Sounds good. I think those were some good ideas and it makes sense to follow up on them. I'll keep poking around for now.

**Comment by apainintheneck** [comment]
Okay, well after looking around I just don't think it's really worth it to pursue the HTTP `HEAD` angle right now.

Essentially, there are a few reasons for this.
1. Very few casks (79/4016 in Homebrew-cask by my count) actually have `version :latest` set so it's seems like premature optimization.
2. It would probably require changing multiple download strategies to get and store additional `last-modified` and `ETag` information during non-upgrade installs which seems like more trouble than it's worth.
3. Like I mentioned before HTTP `HEAD` support doesn't seem to be universal so we'd still have to fallback to the sha anyway.

That being said I did change one thing and you'll have to tell me what you think. Essentially, I decided to store the download path in the Cask object instead of the Installer. This means that the download will only happen once during the lifetime of a Cask object. It passes all of the tests but there might be a scenario I'm not thinking of where you'd need to download a Cask multiple times in one go. It looks like this now.

```ruby
def download(quiet: nil, timeout: nil)
  # Store cask download path in cask to prevent multiple downloads in a row when checking if it's outdated
  @cask.download ||= downloader.fetch(quiet: quiet, verify_download_integrity: @verify_download_integrity,
                                      timeout: timeout)
end
```

Other than that this PR should be ready.

**Comment by MikeMcQuaid** [comment]
> Okay, well after looking around I just don't think it's really worth it to pursue the HTTP `HEAD` angle right now.

Thanks for investigating and the great write-up ❤️ 

> That being said I did change one thing and you'll have to tell me what you think. Essentially, I decided to store the download path in the Cask object instead of the Installer. This means that the download will only happen once during the lifetime of a Cask object. It passes all of the tests but there might be a scenario I'm not thinking of where you'd need to download a Cask multiple times in one go. It looks like this now.

Amazing work!

**Review by MikeMcQuaid — APPROVED** [review]
Amazing work as usual, thanks @apainintheneck!
